platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -506,28 +499,74 @@ private void shrinkIfNeeded(final long newMinValidIndex) {
         }
 
         // clean up a chunk with data
-        final int firstChunkWithData = (int) (newMinValidIndex / numLongsPerChunk);
+        final int firstChunkWithDataIndex = (int) (newMinValidIndex / numLongsPerChunk);
         final long numberOfElementsToCleanUp = (newMinValidIndex % numLongsPerChunk);
-        partialChunkCleanup(firstChunkWithData, numberOfElementsToCleanUp);
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfElementsToCleanUp > 0) {
+            partialChunkCleanup(chunk, true, numberOfElementsToCleanUp);
+        }
 
         // clean up chunk(s) reserved for buffer
-        for (int i = firstValidChunkWithBuffer; i < firstChunkWithData; i++) {
-            partialChunkCleanup(i, numLongsPerChunk);
+        for (int i = firstValidChunkWithBuffer; i < firstChunkWithDataIndex; i++) {
+            chunk = chunkList.get(i);
+            if (chunk != null) {
+                partialChunkCleanup(chunk, true, numLongsPerChunk);
+            }
+        }
+    }
+
+    /**
+     * Deletes values from {@code newMaxValidIndex} to the end of the list and releases memory chunks reserved for these values.
+     *
+     * @param oldMaxValidIndex old maximal valid index, former right boundary of the list
+     * @param newMaxValidIndex new maximal valid index, new right boundary of the list
+     */
+    private void shrinkRightSideIfNeeded(long oldMaxValidIndex, long newMaxValidIndex) {
+        final int listLength = chunkList.length();
+        final int lastValidChunkWithBufferIndex =
+                (int) Math.min((newMaxValidIndex + reservedBufferLength) / numLongsPerChunk, listLength - 1);
+        final int firstChunkIndexToDelete = lastValidChunkWithBufferIndex + 1;
+        final int numberOfChunks = calculateNumberOfChunks(oldMaxValidIndex);
+
+        for (int i = firstChunkIndexToDelete; i < numberOfChunks; i++) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                releaseChunk(chunk);
+            }
+        }
+
+        // clean up a chunk with data
+        final int firstChunkWithDataIndex = (int) (newMaxValidIndex / numLongsPerChunk);
+        final long numberOfEntriesToCleanUp = numLongsPerChunk - (newMaxValidIndex % numLongsPerChunk) - 1;
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfEntriesToCleanUp > 0) {
+            partialChunkCleanup(chunk, false, numberOfEntriesToCleanUp);
+        }
+
+        // clean up chunk(s) reserved for buffer
+        for (int i = firstChunkWithDataIndex + 1; i <= lastValidChunkWithBufferIndex; i++) {
+            chunk = chunkList.get(i);
+            if (chunk != null) {
+                partialChunkCleanup(chunk, false, numLongsPerChunk);
```

Comment:
s/false/true/. `false` should also work, but not consistent with a similar block in `shrinkLeftSideIfNeeded()`.

Reply:
I'd argue it's consistent with the previous  `partialChunkCleanup(chunk, false, numberOfEntriesToCleanUp)` call in this method. You're right - if we clean up the whole chunk it shouldn't matter but `true` in this case looks less confusing because this method is dedicated to the clean up of the _right_ side, and therefore it passes `false` here.